sql: add bitwise operation for varbit with variable length

cty123 opened the pull request:
As raised in #107821, the existing bitwise operations in CRDB follow the same standard as Postgresql that require the two operands of varbit type to have the same length. The restriction makes it easier for the implementation but harder for users. As of today, we need to apply casting and left padding to make sure the operands have the same length. Instead, it might be useful to have CRDB built-in functions that can handle this. Here I have implemented 2 basic bitwise operation functions that are tailored for varbit typed data of arbitrary length.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=107863) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=107863) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=107863) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107863)
<!-- Reviewable:end -->

rharding6373 reviewed (changes requested):
Thank you for implementing the variable-length bitwise operations! This looks really nice, I only have a few minor suggestions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-:-NahZaH-CcLqXUMgpQ_F:b-at91mk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cty123)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-NahYY9z-_Q_O2DrWGiM):*
Could you add the following text to the bottom of your commit message? Thanks!

```
Epic: None
Informs: #107821 

Release Notes(sql): Adds built-in functions `varbit_or_unsigned` and `varbit_and_unsigned` for variable-length input bitwise OR and AND operations, respectively.
```
___
*[`pkg/sql/sem/builtins/builtins.go` line 8193 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-NahXa8L52ldzy6zJbzW:-NahXa8L52ldzy6zJbzX:b6g5jkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/sem/builtins/builtins.go#L8193)):*
> ```Go
> 			},
> 			types.VarBit,
> 			"Calculates bitwise OR value of bit array 'a' and 'b' that may have different lengths.",
> ```

nit: s/array/arrays here and in all overload infos.
___
*[`pkg/sql/sem/builtins/builtins.go` line 8200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-NarWYxS9t3goivMMXm_:-NarWYxS9t3goivMMXma:b5s243h) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/sem/builtins/builtins.go#L8200)):*
> ```Go
> 			},
> 			types.VarBit,
> 			"Calculates bitwise OR value of bit array 'a' and 'b' that may have different lengths.",
> ```

Could you add to the description here and for all other functions that a and b are considered unsigned?
___
*[`pkg/sql/sem/builtins/builtins.go` line 11064 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-Nage8fG0XpmRM2D_JKX:-Nage8fG0XpmRM2D_JKY:b-nxzczt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/sem/builtins/builtins.go#L11064)):*
> ```Go
> }
> 
> // Perform bitwise AND operation 2 bit strings that may have different lengths. The function applies left padding implicitly.
> ```

It's worth mentioning here that the padding is unsigned, i.e., only pads 0s.
___
*[`pkg/sql/sem/builtins/builtins.go` line 11064 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-NahQxEr3KI4fdv5N0mM:-NahQxEr3KI4fdv5N0mN:b7588g4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/sem/builtins/builtins.go#L11064)):*
> ```Go
> }
> 
> // Perform bitwise AND operation 2 bit strings that may have different lengths. The function applies left padding implicitly.
> ```

nit: s/2/to here and below.
___
*[`pkg/sql/sem/builtins/builtins.go` line 11064 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-NahXzAm4kH_WN9lAQEZ:-NahXzAm4kH_WN9lAQE_:bpf6pe) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/sem/builtins/builtins.go#L11064)):*
> ```Go
> }
> 
> // Perform bitwise AND operation 2 bit strings that may have different lengths. The function applies left padding implicitly.
> ```

nit: Do you mind line-wrapping the comment to 80 chars?
___
*[`pkg/sql/logictest/testdata/logic_test/builtin_function` line 4045 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-Nagdtr_3UT_22xRa7E4:-Nagdtra9NTIOfLxuZ2M:b-xi20ab) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/logictest/testdata/logic_test/builtin_function#L4045)):*
> ```
> SELECT varbit_or('1010010', '0101');
> ----
> 1010101
> ```

This should be 1010111, I think.
___
*[`pkg/sql/logictest/testdata/logic_test/builtin_function` line 4091 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-NahRWVV6_JD5RmVl_tx:-NahRWVV6_JD5RmVl_ty:b3xni0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/logictest/testdata/logic_test/builtin_function#L4091)):*
> ```
> SELECT varbit_and('1010010'::varbit, '0101'::varbit);
> ----
> 0000000
> ```

Could you also add the following test cases for edge cases?

```
# Test for invalid inputs.
statement error pgcode 22P02 could not parse string as bit array: "n" is not a valid binary digit
SELECT varbit_or('not binary', '111')

statement error pgcode 22P02 could not parse string as bit array: "n" is not a valid binary digit
SELECT varbit_or('111', 'not binary')

statement error pgcode 22P02 could not parse string as bit array: "n" is not a valid binary digit
SELECT varbit_and('not binary', '111')

statement error pgcode 22P02 could not parse string as bit array: "n" is not a valid binary digit
SELECT varbit_and('111', 'not binary')

# Accept hex as well as binary inputs.
query T
SELECT varbit_or('xfe8c', '1111')
----
1111111010001111

query T
SELECT varbit_and('1111', 'xfe8c')
----
0000000000001100

# Test on a large (>64 bit) input.
query T
SELECT varbit_or('xffffffffffffffffff', '010')
----
111111111111111111111111111111111111111111111111111111111111111111111111
```
___
*[`pkg/sql/sem/builtins/builtins_test.go` line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-NahUe8i3ZAb63iR2LSj:-NahUe8i3ZAb63iR2LSk:b-380un4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/sem/builtins/builtins_test.go#L786)):*
> ```Go
> }
> 
> func TestVarbitOrAnd(t *testing.T) {
> ```

Thank you for the unit test! This is great.


<!-- Sent from Reviewable.io -->

mgartner commented:
I'm a tad concerned that the name of these new builtins, which are not aggregate functions, are similar to names of aggregate functions, `bool_and`, `bool_or`, `bit_and`, `bit_or`, and `bit_xor`. Will that be unnecessarily confusing for users? Are there any sensible alternative names?





<!-- Sent from Reviewable.io -->

rharding6373 commented:
Agree that the names could be improved. Would `bitmask_and` and `bitmask_or` be any better? It's a similar pattern, but since `bitmask` isn't a type maybe that would be a bit more clear.

cty123 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107863)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rharding6373)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-NahYY9z-_Q_O2DrWGiM):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Could you add the following text to the bottom of your commit message? Thanks!

```
Epic: None
Informs: #107821 

Release Notes(sql): Adds built-in functions `varbit_or_unsigned` and `varbit_and_unsigned` for variable-length input bitwise OR and AND operations, respectively.
```
</blockquote></details>

Should this be 

```
... Adds built-in functions `varbit_or` and `varbit_and` for...
```

Because in functions.md the 2 functions I added are ```varbit_or``` and ```varbit_and```. Or are you suggesting I should rename them to ```varbit_or_unsigned``` and ```varbit_and_unsigned```? It would be a bit weird if this is case, because Postgresql and CRDB both have ```bit_or``` and ```bit_and``` functions rather than ```bit_or_unsigned``` or ```bit_and_unsigned```.

Or are they just written this way in the commit message to indicate they are only for unsigned varbit?
___
*[`pkg/sql/sem/builtins/builtins.go` line 8200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-NarWYxS9t3goivMMXm_:-Narecli8z_YDs0L9SJx:b-1ty88n) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/sem/builtins/builtins.go#L8200)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Could you add to the description here and for all other functions that a and b are considered unsigned?
</blockquote></details>

I am fine with this, but just a little bit weird. Why would varbit have signs? I mean, from my perspective, they are just an array of 0s and 1s, and it's really up to the user to translate them, with 2s complement, 1s complement or unsigned integer. At least on Postgresql documentation, I didn't see any emphasis on the operands being unsigned,

https://www.postgresql.org/docs/9.4/functions-bitstring.html
https://www.postgresql.org/docs/9.1/functions-aggregate.html


<!-- Sent from Reviewable.io -->

cty123 commented:
* Renamed ```varbit_or->bitmask_or```, ```varbit_and->bitmask_and```
* Added ```bitmask_xor``` with more testings(I see no reason no to add xor function)
* Fixed the grammar

cty123 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107863)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rharding6373)

___
*[`pkg/sql/sem/builtins/builtins.go` line 11064 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-Nage8fG0XpmRM2D_JKX:-Nas671HBprfEFOhFxNL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/sem/builtins/builtins.go#L11064)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

It's worth mentioning here that the padding is unsigned, i.e., only pads 0s.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/builtin_function` line 4045 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-Nagdtr_3UT_22xRa7E4:-Nas6BSo3ipEhhj4rS0z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/logictest/testdata/logic_test/builtin_function#L4045)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

This should be 1010111, I think.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/builtin_function` line 4091 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-NahRWVV6_JD5RmVl_tx:-Nas6B-J3dsPNEDOL2Tj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/logictest/testdata/logic_test/builtin_function#L4091)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Could you also add the following test cases for edge cases?

```
# Test for invalid inputs.
statement error pgcode 22P02 could not parse string as bit array: "n" is not a valid binary digit
SELECT varbit_or('not binary', '111')

statement error pgcode 22P02 could not parse string as bit array: "n" is not a valid binary digit
SELECT varbit_or('111', 'not binary')

statement error pgcode 22P02 could not parse string as bit array: "n" is not a valid binary digit
SELECT varbit_and('not binary', '111')

statement error pgcode 22P02 could not parse string as bit array: "n" is not a valid binary digit
SELECT varbit_and('111', 'not binary')

# Accept hex as well as binary inputs.
query T
SELECT varbit_or('xfe8c', '1111')
----
1111111010001111

query T
SELECT varbit_and('1111', 'xfe8c')
----
0000000000001100

# Test on a large (>64 bit) input.
query T
SELECT varbit_or('xffffffffffffffffff', '010')
----
111111111111111111111111111111111111111111111111111111111111111111111111
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

cty123 committed:
sql: add bitwise operation for varbit with variable length

As raised in #107821, the existing bitwise operations in CRDB follow the
same standard as Postgresql that require the two operands of varbit type
to have the same length. The restriction makes it easier for the
implementation but harder for users. As of today, we need to apply
casting and left padding to make sure the operands have the same length.
Instead, it might be useful to have CRDB built-in functions that can
handle this. Here I have implemented 3 basic bitwise operation functions
that are tailored for varbit typed data of arbitrary length.

Epic: None
Informs: #107821

Release note (sql change): Adds built-in functions `bitmask_or`,
`bitmask_and` and `bitmask_xor` for variable-length input bitwise OR,
AND, and XOR operations, respectively.

rharding6373 commented:
Thanks for your continued effort on these builtins! I just got back from vacation and will review your most recent changes this week.

rharding6373 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for the changes! I made minor changes to the commit message to reflect the new builtin names and our format requirements. Everything else looks good to me.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-:-Nc3nQxo2u7zNzR7uH4v:b-iecbii)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cty123)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-NahYY9z-_Q_O2DrWGiM):*
<details><summary><i>Previously, cty123 (cty) wrote…</i></summary><blockquote>

Should this be 

```
... Adds built-in functions `varbit_or` and `varbit_and` for...
```

Because in functions.md the 2 functions I added are ```varbit_or``` and ```varbit_and```. Or are you suggesting I should rename them to ```varbit_or_unsigned``` and ```varbit_and_unsigned```? It would be a bit weird if this is case, because Postgresql and CRDB both have ```bit_or``` and ```bit_and``` functions rather than ```bit_or_unsigned``` or ```bit_and_unsigned```.

Or are they just written this way in the commit message to indicate they are only for unsigned varbit?
</blockquote></details>

You're right, I forgot to edit this comment before publishing my initial review. I updated the commit message with the new names.
___
*[`pkg/sql/sem/builtins/builtins.go` line 8200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107863#-NarWYxS9t3goivMMXm_:-NariKEV7Vj7nHUo18gS:b-fkhzyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7028ed112b47fa6319e1b0165e0de3a259a099cf/pkg/sql/sem/builtins/builtins.go#L8200)):*
<details><summary><i>Previously, cty123 (cty) wrote…</i></summary><blockquote>

I am fine with this, but just a little bit weird. Why would varbit have signs? I mean, from my perspective, they are just an array of 0s and 1s, and it's really up to the user to translate them, with 2s complement, 1s complement or unsigned integer. At least on Postgresql documentation, I didn't see any emphasis on the operands being unsigned,

https://www.postgresql.org/docs/9.4/functions-bitstring.html
https://www.postgresql.org/docs/9.1/functions-aggregate.html
</blockquote></details>

Thanks for this change. For context, we had some internal discussion about whether we should support a signed version in the future that pads with `'1'` if the most significant bit of the shorter input is `'1'` instead of `'0'`. The rationale being that even though varbits aren't a signed type themselves, you have to decide how to interpret them once you have to extend the length of one of them in order to do logical bitwise operations on them. Both the postgres builtins and the bit string operators require inputs to be the same length which forces the user to interpret them. 


<!-- Sent from Reviewable.io -->

rharding6373 reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107863)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cty123)



<!-- Sent from Reviewable.io -->

rharding6373 commented:
bors r=rharding6373
